recon/src/recon/pending_cache.rs

```diff
@@ -0,0 +1,153 @@
+use std::collections::HashMap;
+
+use crate::{Key, PendingItem, ReconItem};
+
+#[derive(Debug, Default)]
+/// This struct manages tracking items that we attempted to deliver but could not be processed
+/// because we need to discover additional information.
+pub struct PendingCache<K: Key> {
+    by_needed_key: HashMap<Vec<u8>, ReconItem<K>>,
+    item_keys: std::collections::HashSet<Vec<u8>>,
+    max_size: usize,
+}
+
+impl<K: Key> PendingCache<K> {
+    /// Create a new PendingCache with the given capacity
+    pub fn new(max_size: usize) -> Self {
+        Self {
+            by_needed_key: Default::default(),
+            item_keys: Default::default(),
+            max_size,
+        }
+    }
+
+    fn capacity(&self) -> usize {
+        self.max_size.saturating_sub(self.size())
+    }
+
+    fn size(&self) -> usize {
+        debug_assert_eq!(self.by_needed_key.len(), self.item_keys.len());
+        self.by_needed_key.len()
+    }
+
+    /// Stop tracking an item and return it if the key it the item it required shows up
+    pub fn remove_by_needed(&mut self, new_item: &ReconItem<K>) -> Option<ReconItem<K>> {
+        if let Some(ok_now) = self.by_needed_key.remove(new_item.key.as_bytes()) {
+            self.item_keys.remove(ok_now.key.as_bytes());
+            Some(ok_now)
+        } else {
+            None
+        }
+    }
+
+    /// Returns true if we're already tracking the given item
+    pub fn is_tracking(&self, item: &ReconItem<K>) -> bool {
+        self.item_keys.contains(item.key.as_bytes())
+    }
+
+    /// Update our cache to include the new items up to the allowed capacity.
+    /// Returns a tuple of (items tracked, capacity remaining)
+    pub fn track_pending(&mut self, items: &mut Vec<PendingItem<K>>) -> (usize, usize) {
+        let mut tracked = 0;
+        let mut allowed = self.capacity();
+        if allowed > 0 {
+            while let Some(val) = items.pop() {
+                if !self.item_keys.contains(val.item.key.as_bytes()) {
+                    self.item_keys.insert(val.item.key.as_bytes().to_vec());
+                    self.by_needed_key
+                        .insert(val.required_key.as_bytes().to_vec(), val.item);
+                    allowed = allowed.saturating_sub(1);
+                    tracked += 1;
+                }
+                if allowed == 0 || items.is_empty() {
```

Comment:
can we move these checks into the while loop condition?

Reply:
not with the while let (unstable rust feature) but I could flip the loops so this is the outer one

Reply:
changed to a for loop using drain

Reply:
This has been deleted for now and we can add something to the service when we create these items.